Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs

```diff
@@ -46,9 +46,16 @@ public override void Update()
             {
                 base.Update();
 
-                foreach (var provider in dataProviders)
+                try
                 {
-                    provider.Update();
+                    foreach (var provider in dataProviders)
+                    {
+                        provider.Update();
+                    }
+                }
+                catch(InvalidOperationException)
+                {
+                    Debug.Log("Data provider collection was changed during Update.");
```

Comment:
It feels like `try`/`catch`ing exceptions due to the collection changing instead of more completely managing the lifecycle could lead to unexpected behavior.

For example (but please correct me if I'm wrong here!):

1. In update, the profile changes
1. This exception is thrown, and (as far as I can tell), the new providers don't get `Update` called on them and instead this method bails
1. `LateUpdate` is called on all (now, the new) data providers

If a data provider does something in LateUpdate to follow up on something that's expected to be done in Update, and I think there's a fairly strong promise that system lifecycle happens in a specific order, the provider _could_ get into a weird state.

Reply:
Also it looks like only `Update` and `LateUpdate` are wrapped. What if I try to change the profile when a system is enabled / disabled / reset (maybe not a common scenario, but I'm not sure)?

Reply:
To follow up on this, I remember we had a conversation a while back about more explicitly handling where / when the profiles are torn down and started up. Something like:

1. New profile is set
1. Old services are torn down
1. All remaining service lifecycle code is cancelled for this frame
1. Next frame, new services are spun up
1. Normal service lifecycle resumes

Reply:
All this to say I really think handling the lifecycle of all this more explicitly and completely might lead to not having to spot-fix specific issues one at a time (for example, if we wait until the next frame to try to recreate the UIRaycastCamera, would we have to make the `DestroyImmediate` race condition change? or does the frame boundary complete the clean-up and leave the scene fresh for the new services?).

Reply:
> if we wait until the next frame to try to recreate the UIRaycastCamera, would we have to make the DestroyImmediate race condition change? 

that appears to be an issue based on Unity indicating the object exists after Destroy and before cleanup. MRTK is checking the next frame and, seeing that the object exisst, does not recreate it.

Reply:
Super appreciate the feedback here!

Reply:
> All remaining service lifecycle code is cancelled for this frame

I'll see what i can do to ensure complete update/cleanup between changes. It may introduce an extra frame, or so, between shutdown and restart.

Reply:
> What if I try to change the profile when a system is enabled / disabled / reset.

The [Late]Update try/catch is only for data providers. Services are covered by the handling in ExecuteOnAllServices* methods, which are called in most (very few exceptions) situations.

An earlier incarnation of this change had exception handing for all data provider methods (enable, disable, etc) and that was originally deemed unneccessary.

Happy to revisit!